### ISSUE TITLE
[HOLD for payment 2024-10-30] [$1000] Create a useResponsiveLayout hook

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#30528]
Slack context: https://expensify.slack.com/archives/C01GTK53T8Q/p1698364514493649

### Problem
There are some times when we want to use the same layout on mobile devices _or_ in the RHP on wide screens. In these cases, we can't rely on `isSmallScreenWidth` from `useWindowDimensions` to determine which layout to use, so we have hard-coded and/or different solutions we use to address the same problem.

### Solution
1. In the navigation stack, add an `isInRHP` route prop to all screens in the RHP.
1. Write a custom hook called `useResponsiveLayout` that does something like this:

    ```js
    function useResponsiveLayout() {
        const {isSmallScreenWidth} = useWindowDimensions();
        const {params} = useRoute();
        return {
            shouldUseNarrowLayout = isSmallScreenWidth || params.isInRHP,
        };
    }
    ```
1. Replace most layout-motivated uses of useWindowDimensions with useResponsiveLayout like so:

    ```diff
    - const {isSmallScreenWidth} = useWindowDimensions();
    + const {shouldUseNarrowLayout} = useResponsiveLayout();
    ```
    
1. Introduce a new lint rule that prohibits using `isSmallScreenWidth` from `useWindowDimensions`, advising developers to use `shouldUseNarrowLayout` instead.
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f36dcb081cdf74b2</li>
        <li>Upwork Job ID: 1717949899365408768</li>
        <li>Last Price Increase: 2023-12-29</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @alexpensify</details>

**Comment by alitoshmatov** [comment]
I can work on this if no one is assigned yet, or should I write proposal

**Comment by WebflowGuru** [comment]
**Problem:**
In the current application, there is a challenge when needing to use the same layout for both mobile devices and the right-hand panel (RHP) on wide screens. This problem arises because it's not always feasible to rely on the isSmallScreenWidth variable from useWindowDimensions to determine which layout to use. As a result, developers often resort to hard-coded and inconsistent solutions to address this issue.

**Solution:**
To address this problem, the following changes are proposed:

Route Prop for RHP Detection: In the navigation stack, it is suggested to add an isInRHP route prop to all screens within the RHP. This addition allows for easy identification of screens within the RHP.

Custom Hook - useResponsiveLayout: Create a custom hook called useResponsiveLayout to determine the appropriate layout based on the screen width and the isInRHP route prop.

```
function useResponsiveLayout() {
    const { isSmallScreenWidth } = useWindowDimensions();
    const { params } = useRoute();
    return {
        shouldUseNarrowLayout: isSmallScreenWidth || params.isInRHP,
    };
}
```
Replace useWindowDimensions with useResponsiveLayout: To ensure a consistent approach to layout handling, replace most of the current uses of useWindowDimensions with useResponsiveLayout as shown below:

```
// Before
const { isSmallScreenWidth } = useWindowDimensions();
```

// After
`const { shouldUseNarrowLayout } = useResponsiveLayout();`
Introduce a Lint Rule: To maintain code quality and consistency, introduce a new lint rule that discourages the use of isSmallScreenWidth from useWindowDimensions. Developers are advised to use shouldUseNarrowLayout instead.
This proposed solution aims to provide a more unified and maintainable approach to handling responsive layouts in the application. These changes simplify the management of layouts across various screen sizes and ensure a more consistent user experience.

**Comment by roryabraham** [comment]
Thanks for all the proposals everyone. At this time I've decided to move forward with the one that feels the clearest and most high-quality, which in my opinion is from @rayane-djouah [here](https://github.com/Expensify/App/issues/30528#issuecomment-1783343278).

**Comment by getusha** [comment]
@rayane-djouah any updates?

**Comment by maddylewis** [comment]
can someone provide an update on the status of this issue? thanks!

**Comment by roryabraham** [comment]
Since we're splitting this up into a few PRs, @rayane-djouah, can you lay out the related PRs in sequence please?

**Comment by roryabraham** [comment]
Doubling the bounty here due to large scope

**Comment by maddylewis** [comment]
moving to daily since payment is coming up soon

**Comment by getusha** [comment]
@maddylewis since we are splitting the PR into multiple small PRs we should only process payment once the last PR gets merged. this is only the first one.

**Comment by roryabraham** [comment]
Agreed. Left some comments on the [2nd PR](https://github.com/Expensify/App/pull/33426).

**Comment by roryabraham** [comment]
@rayane-djouah we're awaiting changes in the 2nd PR

**Comment by alexpensify** [comment]
Based on this comment [here](https://github.com/Expensify/App/issues/30528#issuecomment-1883448834), this one is not ready for payment and there is an upcoming PR that needs to go to production before I can move forward with the payment process. 

For now, I updated the Upwork job link: 

https://www.upwork.com/jobs/~0123a7c663aaab8163

@getusha and @rayane-djouah -- please apply and we can be ready for the payment date. 

**Heads up, I will be offline until Tuesday, but will catch up then to identify the status of the last PR. I will not be actively watching over this GitHub during that period. If anything urgent is needed here, please ask for help in the Open Source Slack Room-- thanks!**

**Comment by alexpensify** [comment]
@getusha two requests: 

1. Please apply to the Upwork job: https://www.upwork.com/jobs/~0123a7c663aaab8163

2. Can we get feedback if this Deploy Blocker notice is accurate to the first PR or unrelated to this GH? Thanks!

**Comment by alexpensify** [comment]
@getusha and @roryabraham - to confirm, this one is pending one last PR, so the January 24 payment date is inaccurate - right? Thanks!

**Comment by roryabraham** [comment]
Yes, there's more than 1 outstanding PR. This should not be awaiting payment yet.

@getusha any update on when you'll be able to review this? It's not particularly high-priority, but better to not have it drag on longer than necessary. If you're currently prioritizing something else, that might be ok – just try to communicate a bit more actively. Thanks!

**Comment by alexpensify** [comment]
Thank you for confirming!

**Comment by getusha** [comment]
@alexpensify here is a list of PRs we should complete before payment https://github.com/Expensify/App/issues/30528#issuecomment-1868485991

@rayane-djouah @roryabraham i think we should avoid putting the issue link on upcoming PRs to make the automation stop triggering.

**Comment by alexpensify** [comment]
Thank you for this update and noted!

**Comment by getusha** [comment]
@rayane-djouah when do you think you'll be able to raise the next PR?

**Comment by getusha** [comment]
Thanks, will try to complete the review tomorrow.

**Comment by roryabraham** [comment]
@rayane-djouah @getusha just checking in here – how are we doing and what are our next steps?

**Comment by getusha** [comment]
>needs to update https://github.com/Expensify/App/pull/34837 to resolve some merge conflicts. Anything else?

Yes, @rayane-djouah will complete the checklist & resolve conflicts.

**Comment by alexpensify** [comment]
Looks like we are making progress here!

**Comment by alexpensify** [comment]
Weekly Update: Making progress

**Comment by alexpensify** [comment]
Weekly Update: @rayane-djouah some of the PRs are still in draft mode 

Can we get an update in the PR list you put together here - https://github.com/Expensify/App/issues/30528#issuecomment-1868485991

**Comment by alexpensify** [comment]
The PR listed above is still in review

**Comment by alexpensify** [comment]
Weekly Update: the PR is moving forward.

**Comment by alexpensify** [comment]
Weekly Update: The PR is going through reviews but moving along.

**Comment by alexpensify** [comment]
Weekly Update: The PR is in review to make sure testing passes.

**Comment by alexpensify** [comment]
Weekly Update: The PR went through another review process and more testing.

**Comment by alexpensify** [comment]
Weekly Update: More discussion in the PR due to recent findings

**Comment by alexpensify** [comment]
@roryabraham can I get feedback regarding which Wave or VIP to associate this one with? -- thanks!

**Comment by alexpensify** [comment]
Weekly Update: We are in a focus period, I'll check in next week.

**Comment by roryabraham** [comment]
This is a pretty old issue, and I'm not sure which wave or VIP I'd associate it with. It's basically a low-level code improvement to help avoid a semi-common mistake and make the code more consistent. Some recent discussion about it in slack [here](https://expensify.slack.com/archives/C03TME82F/p1714085494359619).

**Comment by alexpensify** [comment]
Thanks for the context @roryabraham! 

Weekly Update: The PR is moving in review again: https://github.com/Expensify/App/pull/36292

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Tuesday, May 28, 2024, and will not actively watch over this GitHub during that period.

If anything urgent is needed here, please ask for help in the #expensify-open-source Slack Room-- thanks!

**Comment by alexpensify** [comment]
Catching up from being OOO, it looks like this one was on hold due to the freeze but we can carry on now.

**Comment by roryabraham** [comment]
PR merged today

**Comment by alexpensify** [comment]
Weekly Update: It looks like there was regression, but that has been addressed. I'll check again next week to confirm if we can move forward with the payment process.

**Comment by alexpensify** [comment]
Automation failed here, I need to start the payment process tomorrow since today marks 7 days in production.

**Comment by getusha** [comment]
@alexpensify there are still other PRs we need to get merged https://github.com/Expensify/App/issues/30528#issuecomment-1868485991

**Comment by alexpensify** [comment]
@getusha, thanks for flagging; I appreciate it. With that said, are we still striking the completed ones? If not, we should to keep a better record of what's left. For now, I'll hold on the payment process.

**Comment by getusha** [comment]
@rayane-djouah i think we can put the remaining changes in a single PR except the `eslint-config-expensify` changes

**Comment by roryabraham** [comment]
sounds good 👍🏼

**Comment by alexpensify** [comment]
@getusha and @rayane-djouah - can either of you confirm if this notice is accurate or not? Thanks!

**Comment by alexpensify** [comment]
Ok, waiting on https://github.com/Expensify/eslint-config-expensify/pull/116 to go into production

**Comment by roryabraham** [comment]
pulling in @mountiny to take over while I'm on parental leave

**Comment by mountiny** [comment]
ESLint rule merged @rayane-djouah can you raise App PR now please?

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Tuesday, October 22, 2024, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by mountiny** [comment]
No regression tests required this is a tool for developers

The payment summary is $1000 to @rayane-djouah and to @getusha

**Comment by alexpensify** [comment]
Payouts due: 2024-10-30

- [ ] Contributor: $250 @rayane-djouah 
- [ ] Reviewer: $250 @getusha 

Upwork job is [here](https://www.upwork.com/jobs/~021851827931519059847). I had to create a new job in Upwork. Please accept and I can complete the payment process. Thanks!

**Comment by alexpensify** [comment]
🤦🏼 good catch - Upwork was right. I've updated the summary here and paid @rayane-djouah via Upwork. I'm waiting for @getusha to accept.

**Comment by alexpensify** [comment]
Status: Waiting for @getusha to accept in Upwork

**Comment by alexpensify** [comment]
Thanks for flagging! This GH is from 2023. I think it needs to be paid via Chat, but let me double-check.

**Comment by alexpensify** [comment]
I confirmed that this one is from 2023. It will need to be paid via Upwork. I will move this to weekly since I'm OOO until Tuesday. @getusha, keep me posted when you accept it in Upwork. I'll complete the process when I'm back online.

**Comment by alexpensify** [comment]
All set, everyone has been paid via Upwork: https://github.com/Expensify/App/issues/30528#issuecomment-2448951102
